### PR TITLE
Improved speed of lp forward pass

### PR DESCRIPTION
**Pull request by abh2k**

### CODE REVIEW

**Comment by abh2k** [comment]
The idea here is pretty much similar to the forward pass of mean_forward.

**Review by zoq — COMMENTED** [review]
I guess you did time this against the previous implementaiton?

**Comment by Abhinav-Aidash** [comment]
> I guess you did time this against the previous implementation?

Yes, you can find it here https://colab.research.google.com/drive/1MmJoeJfFaJ--ugz8VAViHsHTjssYCn3L?usp=sharing

**Comment by abh2k** [comment]
Yes, you can find it here https://colab.research.google.com/drive/1MmJoeJfFaJ--ugz8VAViHsHTjssYCn3L?usp=sharing

File: src/mlpack/methods/ann/layer/lp_pooling.hpp [hunk 1]
```diff
@@ -175,14 +185,20 @@ class LpPooling
...
+        }
+        if (colidx >= 1)
+          val -= inputPre(rowEnd, colidx - 1);
 
-        output(i, j) = pow(arma::accu(arma::pow(subInput,
-            normType)), 1.0 / normType);
+        output(i, j) = val;
       }
     }
+    output = arma::pow(output, 1.0 / normType);
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer/lp_pooling.hpp:201]:
```suggestion
    }

    output = arma::pow(output, 1.0 / normType);
```

File: src/mlpack/methods/ann/layer/lp_pooling.hpp [hunk 2]
```diff
@@ -175,14 +185,20 @@ class LpPooling
...
-            arma::span(rowidx, rowEnd),
-            arma::span(colidx, colEnd));
+        val += inputPre(rowEnd, colEnd);
+        if (rowidx >= 1)
+        {
+          if (colidx >= 1)
+            val += inputPre(rowidx - 1, colidx - 1);
+          val -= inputPre(rowidx - 1, colEnd);
+        }
+        if (colidx >= 1)
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer/lp_pooling.hpp:195]:
```suggestion
        }

        if (colidx >= 1)
```

**Review by kartikdutt18 — COMMENTED** [review]
Minor style fixes. Looks good to me!

**Review by kartikdutt18 — APPROVED** [review]
LGTM!

**Review by zoq — APPROVED** [review]
Looks great to me, thanks for putting this together.

**Comment by zoq** [comment]
Thanks those performance improvements are great.

**Pull request merged**